client-runtime/src/main/java/com/microsoft/rest/ServiceFuture.java

```diff
@@ -43,7 +44,7 @@ protected ServiceFuture() {
             .subscribe(new Action1<ServiceResponse<T>>() {
                 @Override
                 public void call(ServiceResponse<T> t) {
-                    serviceFuture.set(t.body());
+                    serviceFuture.valueSet = serviceFuture.set(t.body());
```

Comment:
should we make the `valueSet` flag a volatile? In case `isCancelled` gets called from a thread other than the subscribe thread, in such case, there is a possibility that `isCancelled` calling thread won't see the new flag value immediately.